File: python/bulk/bulk_import_orders.py
# Bulk import orders - end-to-end lifecycle.
#
# The bulk API is a throughput multiplier for the per-record REST endpoints: you
# upload an NDJSON file of records, submit one job, then poll it to completion.
#
# Flow demonstrated here:
#   1. Build an NDJSON payload (one order per line, each with a stable _merchant_record_id).
#   2. Ask the Bulk API for a presigned S3 upload URL.
#   3. PUT the NDJSON bytes straight to that URL (this step does NOT go through the SDK).
#   4. Submit the job. Orders are insert-only; operation defaults to "insert".
#   5. Poll the job until it reaches a terminal status.
#   6. Print the counts and list any failed records.
#
# For create-or-update semantics, see bulk_upsert_customers.py (customer upsert).

import json
import time
import urllib.request

from ultracart.apis import BulkApi
from ultracart.models import BulkJobRequest
from ultracart.rest import ApiException
from samples import api_client

OBJECT_TYPE = 'order'

# Terminal job statuses - stop polling once the job reaches one of these.
TERMINAL_STATUSES = {'succeeded', 'partial_success', 'failed', 'cancelled'}

# --- 1. Build the NDJSON payload -------------------------------------------
# Each line is exactly what POST /rest/v2/order accepts, PLUS a top-level
# _merchant_record_id - the bulk-surface dedupe key. Use a STABLE value derived
# from your source system (e.g. the legacy order id), NOT a fresh UUID, so that
# re-running a fixed file collapses to the original landing instead of re-inserting.
#
# Build the lines as PLAIN dicts, not SDK model instances: _merchant_record_id is
# a bulk-surface marker, not a field on the per-record Order model.
orders = [
    {
        '_merchant_record_id': 'legacy-order-1029384',
        'merchant_order_id': 'ORD-1029384',
        'creation_dts': '2023-11-14T09:22:00Z',
        'currency_code': 'USD',
        'total': 79.50,
        'customer_profile': {'email': 'jane@example.com'},
        'items': [
            {'merchant_item_id': 'WIDGET-RED', 'quantity': 2,
             'unit_cost': {'currency_code': 'USD', 'value': 35.00}}
        ],
        'shipping': {
            'ship_to': {
                'first_name': 'Jane', 'last_name': 'Doe', 'address1': '123 Main St',
                'city': 'Austin', 'state_region': 'TX', 'postal_code': '78701', 'country_code': 'US'
            },
            'shipping_method': 'Standard'
        }
    },
    {
        '_merchant_record_id': 'legacy-order-1029385',
        'merchant_order_id': 'ORD-1029385',
        'creation_dts': '2023-11-15T14:03:00Z',
        'currency_code': 'USD',
        'total': 35.00,
        'customer_profile': {'email': 'john@example.com'},
        'items': [
            {'merchant_item_id': 'WIDGET-BLUE', 'quantity': 1,
             'unit_cost': {'currency_code': 'USD', 'value': 35.00}}
        ]
    }
]

# NDJSON = one minified JSON object per line, LF-separated, UTF-8. Do NOT wrap the
# records in a JSON array and do NOT pretty-print - the worker parses the file line
# by line. json.dumps with compact separators keeps each record on a single line.
ndjson = '\n'.join(json.dumps(order, separators=(',', ':')) for order in orders)


def main():
    bulk_api = BulkApi(api_client())

    # --- 2. Get a presigned upload URL --------------------------------------
    upload_url_response = bulk_api.bulk_generate_upload_url(OBJECT_TYPE)
    upload_url = upload_url_response.upload_url
    s3_key = upload_url_response.s3_key
    print(f"Upload URL issued (s3_key={s3_key}, "
          f"max_records={upload_url_response.get('max_records', '?')}).")

    # --- 3. Upload the NDJSON straight to the presigned URL ------------------
    # A plain HTTP PUT to S3 using the standard library; this does NOT go through
    # the SDK (the presigned URL already carries its own auth).
    body = ndjson.encode('utf-8')
    put_request = urllib.request.Request(
        upload_url, data=body, method='PUT',
        headers={'Content-Type': 'application/x-ndjson'})
    with urllib.request.urlopen(put_request) as put_response:
        if put_response.status not in (200, 201):
            raise RuntimeError(f"Upload PUT failed: {put_response.status} {put_response.reason}")
    print(f"Uploaded {len(orders)} records ({len(body)} bytes).")

    # --- 4. Submit the job --------------------------------------------------
    # Orders are insert-only; operation defaults to "insert" but is shown for clarity.
    request = BulkJobRequest(s3_key=s3_key, operation='insert')
    # request.webhook_url = 'https://example.com/hooks/bulk'  # optional one-shot completion POST

    submit_response = bulk_api.bulk_submit_job(OBJECT_TYPE, request)
    job = submit_response.bulk_job
    job_id = job.job_id
    print(f"Submitted job {job_id} (status={job.status}).")

    # --- 5. Poll until terminal ---------------------------------------------
    # Bulk endpoints sit behind their own rate-limit bucket, so polling does not
    # consume your normal REST budget - but still poll politely and bound the loop.
    for _ in range(20):
        if job.get('status') in TERMINAL_STATUSES:
            break
        time.sleep(3)
        job = bulk_api.bulk_get_job(OBJECT_TYPE, job_id).bulk_job
        print(f"  status={job.get('status')} "
              f"processed={job.get('processed_records', 0)}/{job.get('total_records', '?')}")

    # --- 6. Report ----------------------------------------------------------
    print(f"\nJob {job_id} finished: {job.get('status')}")
    print(f"  success={job.get('success_count', 0)} "
          f"failed={job.get('fail_count', 0)} duplicate={job.get('duplicate_count', 0)}")

    if job.get('fail_count', 0) > 0:
        print('\nFailed records:')
        cursor = None
        while True:
            kwargs = {'status': 'failed', 'limit': 100}
            if cursor:
                kwargs['cursor'] = cursor
            records_response = bulk_api.bulk_get_job_records(OBJECT_TYPE, job_id, **kwargs)
            for record in records_response.records:
                print(f"  line {record.get('line_number')} "
                      f"[{record.get('merchant_record_id')}] "
                      f"{record.get('error_code')}: {record.get('error_message')}")
            cursor = records_response.get('next_cursor')
            if not cursor:
                break


if __name__ == '__main__':
    try:
        main()
    except ApiException as e:
        # ApiException's __str__ includes the HTTP status and response body.
        print('Bulk import failed:')
        print(e)
        import sys
        sys.exit(1)

File: csharp/bulk/BulkImportOrders.cs
// Bulk import orders — end-to-end lifecycle.
//
// The bulk API is a throughput multiplier for the per-record REST endpoints: you
// upload an NDJSON file of records, submit one job, then poll it to completion.
//
// Flow demonstrated here:
//   1. Build an NDJSON payload (one order per line, each with a stable _merchant_record_id).
//   2. Ask the Bulk API for a presigned S3 upload URL.
//   3. PUT the NDJSON bytes straight to that URL (this step does NOT go through the SDK).
//   4. Submit the job. Orders are insert-only; operation defaults to "insert".
//   5. Poll the job until it reaches a terminal status.
//   6. Print the counts and list any failed records.
//
// For create-or-update semantics, see BulkUpsertCustomers.cs (customer upsert).

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading;
using com.ultracart.admin.v2.Api;
using com.ultracart.admin.v2.Client;
using com.ultracart.admin.v2.Model;
using Newtonsoft.Json;

namespace SdkSample.bulk
{
    // ReSharper disable once ClassNeverInstantiated.Global
    public class BulkImportOrders
    {
        private const string ObjectType = "order";

        // Terminal job statuses — poll until the job reports one of these. Status is a
        // generated enum on the SDK model (Partialsuccess == the wire's "partial_success").
        private static readonly BulkJob.StatusEnum[] TerminalStatuses =
        {
            BulkJob.StatusEnum.Succeeded,
            BulkJob.StatusEnum.Partialsuccess,
            BulkJob.StatusEnum.Failed,
            BulkJob.StatusEnum.Cancelled
        };

        // The S3 upload is a raw HTTP PUT, so it does not go through the SDK.
        private static readonly HttpClient HttpClient = new HttpClient();

        public static void Execute()
        {
            try
            {
                RunBulkImport();
            }
            catch (ApiException e)
            {
                Console.WriteLine("An ApiException occurred.  Please review the following error:");
                Console.WriteLine(e); // <-- change_me: handle gracefully
                Environment.Exit(1);
            }
        }

        private static void RunBulkImport()
        {
            BulkApi bulkApi = Samples.GetBulkApi();

            // --- 1. Build the NDJSON payload -----------------------------------------
            // Each line is exactly what POST /rest/v2/order accepts, PLUS a top-level
            // _merchant_record_id — the bulk-surface dedupe key. Use a STABLE value derived
            // from your source system (e.g. the legacy order id), NOT a fresh UUID, so that
            // re-running a fixed file collapses to the original landing instead of re-inserting.
            //
            // Build the lines as plain objects (anonymous types here), NOT SDK models: the
            // _merchant_record_id marker is a bulk-surface field, not part of the order model.
            var orders = new List<object>
            {
                new
                {
                    _merchant_record_id = "legacy-order-1029384",
                    merchant_order_id = "ORD-1029384",
                    creation_dts = "2023-11-14T09:22:00Z",
                    currency_code = "USD",
                    total = 79.50,
                    customer_profile = new { email = "jane@example.com" },
                    items = new[]
                    {
                        new { merchant_item_id = "WIDGET-RED", quantity = 2, unit_cost = new { currency_code = "USD", value = 35.00 } }
                    },
                    shipping = new
                    {
                        ship_to = new
                        {
                            first_name = "Jane", last_name = "Doe", address1 = "123 Main St",
                            city = "Austin", state_region = "TX", postal_code = "78701", country_code = "US"
                        },
                        shipping_method = "Standard"
                    }
                },
                new
                {
                    _merchant_record_id = "legacy-order-1029385",
                    merchant_order_id = "ORD-1029385",
                    creation_dts = "2023-11-15T14:03:00Z",
                    currency_code = "USD",
                    total = 35.00,
                    customer_profile = new { email = "john@example.com" },
                    items = new[]
                    {
                        new { merchant_item_id = "WIDGET-BLUE", quantity = 1, unit_cost = new { currency_code = "USD", value = 35.00 } }
                    }
                }
            };

            // NDJSON = one minified JSON object per line, LF-separated. Do NOT wrap the records
            // in a JSON array and do NOT pretty-print — the worker parses the file line by line.
            var lines = new List<string>();
            foreach (var order in orders)
            {
                lines.Add(JsonConvert.SerializeObject(order));
            }
            string ndjson = string.Join("\n", lines);
            byte[] payloadBytes = Encoding.UTF8.GetBytes(ndjson);

            // --- 2. Get a presigned upload URL ---------------------------------------
            BulkUploadUrlResponse uploadUrlResponse = bulkApi.BulkGenerateUploadUrl(ObjectType);
            string uploadUrl = uploadUrlResponse.UploadUrl;
            string s3Key = uploadUrlResponse.S3Key;
            Console.WriteLine($"Upload URL issued (s3_key={s3Key}, max_records={uploadUrlResponse.MaxRecords}).");

            // --- 3. Upload the NDJSON straight to the presigned URL ------------------
            // A plain HTTP PUT to S3, using the framework HttpClient rather than the SDK.
            var content = new ByteArrayContent(payloadBytes);
            content.Headers.ContentType = new MediaTypeHeaderValue("application/x-ndjson");
            HttpResponseMessage putResponse = HttpClient.PutAsync(uploadUrl, content).GetAwaiter().GetResult();
            if (!putResponse.IsSuccessStatusCode)
            {
                throw new Exception($"Upload PUT failed: {(int)putResponse.StatusCode} {putResponse.ReasonPhrase}");
            }
            Console.WriteLine($"Uploaded {orders.Count} records ({payloadBytes.Length} bytes).");

            // --- 4. Submit the job --------------------------------------------------
            var request = new BulkJobRequest
            {
                S3Key = s3Key,
                Operation = BulkJobRequest.OperationEnum.Insert // default for orders; shown here for clarity
                // WebhookUrl = "https://example.com/hooks/bulk" // optional one-shot completion POST
            };

            BulkJobResponse submitResponse = bulkApi.BulkSubmitJob(ObjectType, request);
            BulkJob job = submitResponse.BulkJob;
            string jobId = job.JobId;
            Console.WriteLine($"Submitted job {jobId} (status={job.Status}).");

            // --- 5. Poll until terminal (bounded so it can't hang) ------------------
            job = PollUntilTerminal(bulkApi, jobId, job);

            // --- 6. Report ----------------------------------------------------------
            Console.WriteLine($"\nJob {jobId} finished: {job.Status}");
            Console.WriteLine($"  success={job.SuccessCount} failed={job.FailCount} duplicate={job.DuplicateCount}");

            if (job.FailCount > 0)
            {
                Console.WriteLine("\nFailed records:");
                string cursor = null;
                do
                {
                    BulkRecordsResponse recordsResponse = bulkApi.BulkGetJobRecords(ObjectType, jobId, "failed", cursor, 100);
                    foreach (BulkRecord record in recordsResponse.Records)
                    {
                        Console.WriteLine($"  line {record.LineNumber} [{record.MerchantRecordId}] {record.ErrorCode}: {record.ErrorMessage}");
                    }
                    cursor = recordsResponse.NextCursor;
                } while (!string.IsNullOrEmpty(cursor));
            }
        }

        // Poll the job until it reaches a terminal status. Bulk endpoints sit behind their own
        // rate-limit bucket, so polling does not consume your normal REST budget — but still poll
        // politely, and bound the loop (~20 tries) so a stuck job can't hang the sample.
        private static BulkJob PollUntilTerminal(BulkApi bulkApi, string jobId, BulkJob job)
        {
            const int maxTries = 20;
            for (int i = 0; i < maxTries && !IsTerminal(job.Status); i++)
            {
                Thread.Sleep(3000);
                job = bulkApi.BulkGetJob(ObjectType, jobId).BulkJob;
                Console.WriteLine($"  status={job.Status} processed={job.ProcessedRecords}/{job.TotalRecords}");
            }
            return job;
        }

        private static bool IsTerminal(BulkJob.StatusEnum? status)
        {
            return status.HasValue && Array.IndexOf(TerminalStatuses, status.Value) >= 0;
        }

        // C# projects allow a single Main entry point, which already lives in EntryPoint.cs.
        // To run just this sample standalone, comment out EntryPoint.Main and uncomment:
        //
        // public static void Main(string[] args)
        // {
        //     Execute();
        // }
    }
}
